### PR TITLE
fix(vite-plugin-angular): transpile skipped Angular files with TypeScript in Vitest

### LINKED ISSUE #2555: Coverage collection fails with PARSE_ERROR on unused Angular files
### Please provide the environment you discovered this bug in.

- @analogjs/vite-plugin-angular: 2.7.x
- @analogjs/vitest-angular: 2.7.x
- vitest: 4.1.11, @vitest/coverage-v8: 4.1.11
- Angular: 22.1.5, Node: 24.20.0

### Which area/package is the issue in?

vite-plugin-angular

### Description

With V8 coverage enabled, an Angular component that no test imports is omitted from the coverage report.
 
Analog warns that the component is outside the TypeScript program. Coverage collection then logs a `PARSE_ERROR` on TypeScript syntax such as `implements OnInit` or a typed property. The tests still pass.

### Please provide the exception or error you saw

```true
[vite] (client) warning: [@analogjs/vite-plugin-angular]: "<path>/untested.component.ts" contains Angular decorators but is not in the TypeScript program. Ensure it is included in your tsconfig.
  Plugin: @analogjs/vite-plugin-angular
  File: <path>/untested.component.ts
Failed to parse file:///<path>/untested.component.ts. Excluding it from coverage.
 Error [RolldownError]: Parse failed with 2 errors:
'implements' clauses can only be used in TypeScript files.
...
export class UntestedComponent implements OnInit {
                                    ^
Expected `;` but found `:`
...
  production: boolean = environment.production;
                ^
...
    at V8CoverageProvider.remapCoverage (.../@vitest/coverage-v8/dist/provider.js:133:10)
    at V8CoverageProvider.getCoverageMapForUncoveredFiles (.../@vitest/coverage-v8/dist/provider.js:109:4)
    at V8CoverageProvider.generateCoverage (.../@vitest/coverage-v8/dist/provider.js:56:29)
  code: 'PARSE_ERROR', id: undefined, pos: 1662
```

### Other information

This is caught inside coverage-v8's own `remapCoverage` (see `node_modules/@vitest/coverage-v8/dist/provider.js`:130-137 in 4.1.11) — it returns {} and moves on, so it doesn't fail the overall test run.

### I would be willing to submit a PR to fix this issue

- [ ] Yes
- [x] No

### PR DESCRIPTION
**Pull request by mnkprs**
## PR Checklist

Closes #2555

## Affected scope

- Primary scope: vite-plugin-angular
- Secondary scopes: none

## Recommended merge strategy for maintainer [optional]

- [x] Squash merge
- [ ] Rebase merge
- [ ] Other

## What is the new behavior?

`@vitest/coverage-v8` transforms every file matching `coverage.include` to report it as
uncovered, even ones no test ever imports. A component outside the TypeScript program (e.g.
one no spec imports) was returned untranspiled by the main plugin, with a warning, and fell
through to `angularVitestSourcemapPlugin`'s OXC/esbuild path — which either throws on
parameter decorators or emits an `@oxc-project/runtime` import that isn't a project
dependency. When that transform throws or its output fails to parse, coverage-v8 falls back
to the raw `.ts` source and drops the file, producing the `PARSE_ERROR`/`RolldownError`/
`RollupError` in #2555.

Files with any decorator (detected by walking the real AST, so an aliased import,
namespace-qualified reference, or bare decorator are all covered) are now transpiled with
`ts.transpileModule` instead, using the compiler options actually resolved for the active
compilation strategy — including `fastCompile` (the default under `isTest`), which now also
watches its tsconfig `extends` chain and reacts to atomic-save edits so this stays correct
across a live Vitest run. Files with no decorator, and decorated `.tsx` files (whose JSX
lowering mode this fallback can't safely infer), keep the existing OXC/esbuild path
unchanged.

Neither `experimentalDecorators` nor `useDefineForClassFields` is overridden, since this
plugin also runs for files a real test imports and executes, not just ones coverage merely
inspects — a project using standard TC39 decorators keeps their real runtime semantics, and
a class field keeps its real "define"/"assign" behavior. `.mtsx`/`.ctsx` JSX files (both the
OXC and Vite‑6/7 esbuild fallback paths) and `.cts` files without their own CommonJS-only
syntax are also fixed, since they hit adjacent, closely related gaps in the same code path.

A file belonging to a different real npm/pnpm/yarn package than the app (a workspace library
with its own `package.json`, detected by nearest-ancestor lookup rather than directory
nesting or `node_modules`-path matching, both of which break under common monorepo/symlink
layouts) is transpiled with *that* package's own nearest `tsconfig.json` instead of the app's,
so it isn't silently given the app's decorator model.

## Test plan

- [x] `nx format:check` (`pnpm exec prettier --check` on all touched files)
- [x] `pnpm build` (`pnpm nx build vite-plugin-angular`)
- [x] `pnpm test` (`pnpm nx test vite-plugin-angular` — 812 tests passing; plus
      `pnpm exec tsc -p packages/vite-plugin-angular/tsconfig.lib.json --noEmit` and
      `pnpm nx eslint:lint vite-plugin-angular`, both clean of new issues)
- [x] Manual verification: reproduced all three failure modes from #2555 (Vite 8 + OXC with
      `experimentalDecorators`, Vite 8 + OXC without it, Vite 6/7 + esbuild) against a
      standalone repro project built from the built package, confirmed `vitest run --coverage`
      lists every source file with no `PARSE_ERROR`/`RolldownError`/`RollupError`, both with
      and without `experimentalDecorators`.

New/updated specs (fail against the pre-fix code, pass after) live in
`angular-vitest-plugin.spec.ts`, `angular-vite-plugin.spec.ts`, and `utils/plugin-config.spec.ts`.

## Does this PR introduce a breaking change?

- [ ] Yes
- [x] No

## Other information

The same code path is unchanged on `alpha` (v3) — this fix targets `beta` per the repo's
default branch; a maintainer may want to forward-port it.

### CODE REVIEW

**Comment by brandonroberts** [comment]
I suggest using `@analogjs/vite-plugin-angular`’s existing `tsconfig` hook to select a test config that includes all coverage source files. That should bring unimported components into the Angular program and let the existing compiler handle them, avoiding the additional transpilation and config-resolution logic.

File: packages/vite-plugin-angular/src/lib/angular-vitest-plugin.ts [hunk 1]
```diff
@@ -123,14 +397,100 @@ export function angularVitestSourcemapPlugin(
...
+      // nearest `tsconfig.json` (bounded to its own package directory,
+      // so the search can't walk past it and pick up the app's tsconfig
+      // by accident) is read and used instead, so this fallback still
+      // lowers its decorators correctly, just with settings that are
+      // actually this other package's own rather than the app's.
+      const filePackageRoot = findNearestPackageJson(dirname(bareId));
+      const appPackageRoot = getAppPackageRoot?.();
+      const isInDifferentPackage =
+        !!appPackageRoot &&
+        normalizePath(filePackageRoot ?? '') !== normalizePath(appPackageRoot);
```
[the remark this answers is not in the document]
Comment (mnkprs) [reply packages/vite-plugin-angular/src/lib/angular-vitest-plugin.ts]: Good catch — fixed in 30285007: a missing `filePackageRoot` is now treated as "same package as the app" rather than "different", since there's no evidence it's foreign. Added a regression test (`still applies this app's compiler options to a file with no ancestor package.json at all`) that fails against the prior code with exactly the symptom you described (silently dropping `experimentalDecorators: true` to standard decorators).

**Comment by mnkprs** [comment]
Addressed CodeRabbit's functional-correctness finding in 30285007 (a file with no ancestor `package.json` is now treated as this app's own, with a regression test).

Two other CodeRabbit suggestions I'm leaving out of this PR:

- **`typescript` as a peer dependency** (`utils/plugin-config.ts`): this package already imports `typescript` at runtime in `angular-vite-plugin.ts` (pre-dating this PR), and it isn't declared as a dependency/peerDependency there either — this diff just adds two more files that also import it. Since it's a pre-existing gap unrelated to #2555, I'd rather not fold an unrelated packaging change into this fix; happy to open a separate PR for it if useful.
- **Caching the package/tsconfig lookups on the decorator fallback path**: a real perf nitpick, but adding cache invalidation here would need to hook into `onTsconfigChanged` too (as CodeRabbit notes), and I'd rather not add that surface without a concrete case where the extra `existsSync` walk/tsconfig parse per untested decorated file is actually showing up as a bottleneck.

Also, the failing `blog-app-e2e:e2e` check is a runner port conflict (`http://localhost:3000 is already used`) — Nx Cloud's own self-healing bot flagged it as unrelated to this PR.

**Comment by mnkprs** [comment]
@brandonroberts You're right, and I tested it. Using the same `@Injectable`/`@Inject()` pattern from the original report, against the unmodified plugin code: widening the test tsconfig's `include`/`files` to cover the untested file (instead of narrowing to `*.spec.ts` only) brings it into the real Angular/TypeScript program, and the existing compiler handles it correctly with zero warnings. No plugin change needed at all.

Closing this PR — it duplicates a problem that already has a real fix. I'll follow up separately against the scaffolded `tsconfig.spec.json` templates instead: both `create-analog` and the Nx app generator default to `include: ["src/**/*.spec.ts", "src/**/*.d.ts"]`, which is what makes this the default experience rather than a corner case, plus a note in the Vitest testing docs about keeping `coverage.include` and the test tsconfig's `include` in sync. Thanks for the redirect.
